File: backend/models/complaint.js
import mongoose from "mongoose";

const complaintSchema = new mongoose.Schema(
  {
    title: { type: String, required: true, trim: true },
    description: { type: String, required: true, trim: true },
    category: {
      type: String,
      enum: ["Mess", "Hostel Infrastructure", "Academic", "Other"],
      default: "Other",
    },
    status: {
      type: String,
      enum: ["Pending", "In Progress", "Resolved", "Rejected"],
      default: "Pending",
    },
    createdBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Student", // Assuming mainly students complain
        required: true
    },
    createdByName: { type: String }, // Denormalized for easier display
    upvotes: [{ type: mongoose.Schema.Types.ObjectId, ref: "Student" }],
    imageURI: { type: String, trim: true },
  },
  { timestamps: true }
);

const Complaint = mongoose.model("Complaint", complaintSchema);
export default Complaint;

File: frontend/lib/models/complaint.dart
class Complaint {
  final String id;
  final String title;
  final String description;
  final String category;
  final String status;
  final String createdBy;
  final String? createdByName;
  final List<String> upvotes;
  final String? imageURI;
  final Map<String, dynamic>? resolvedBy;
  final String? resolutionNote;
  final DateTime createdAt;


  Complaint({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.status,
    required this.createdBy,
    this.createdByName,
    required this.upvotes,
    this.imageURI,
    this.resolvedBy,
    this.resolutionNote,
    required this.createdAt,

  });

  factory Complaint.fromJson(Map<String, dynamic> json) {
    return Complaint(
      id: json['_id'],
      title: json['title'],
      description: json['description'],
      category: json['category'],
      status: json['status'],
      createdBy: json['createdBy'],
      createdByName: json['createdByName'],
      upvotes: (json['upvotes'] as List).map((e) => e.toString()).toList(),
      imageURI: json['imageURI'],
      resolvedBy: json['resolvedBy'],
      resolutionNote: json['resolutionNote'],
      createdAt: DateTime.parse(json['createdAt']),

    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'category': category,
      'imageURI': imageURI,
    };
  }
}
